src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,396 @@
+                _readHead = _commitHead;
+                returnEnd = _commitHead;
+                _readIndex = 0;
+            }
+            else if (consumedIndex == returnEnd.Length)
+            {
+                var nextBlock = returnEnd.NextSegment;
+                _readHead = nextBlock;
+                _readIndex = 0;
+                returnEnd = nextBlock;
+            }
+            else
+            {
+                _readHead = consumedSegment;
+                _readIndex = consumedIndex;
+            }
+
+            // Remove all blocks that are freed (except the last one)
+            while (returnStart != null && returnStart != returnEnd)
+            {
+                returnStart.ResetMemory();
+                returnStart = returnStart.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void CancelPendingRead()
+        {
+            InternalTokenSource.Cancel();
+        }
+
+        /// <inheritdoc />
+        public override void Complete(Exception exception = null)
+        {
+            if (_isCompleted)
+            {
+                return;
+            }
+
+            _isCompleted = true;
+            if (exception != null)
+            {
+                _exceptionInfo = ExceptionDispatchInfo.Capture(exception);
+            }
+
+            var segment = _readHead;
+            while (segment != null)
+            {
+                segment.ResetMemory();
+                segment = segment.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnWriterCompleted(Action<Exception, object> callback, object state)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            var tokenSource = InternalTokenSource;
+            if (TryReadInternal(tokenSource, out var readResult))
+            {
+                return readResult;
+            }
+
+            var reg = new CancellationTokenRegistration();
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.Register(state => ((StreamPipeReader)state).Cancel(), this);
+            }
+
+            using (reg)
+            {
+                var isCanceled = false;
+                try
+                {
+                    AllocateCommitHead();
+#if NETCOREAPP2_2
+                    var length = await _readingStream.ReadAsync(_commitHead.AvailableMemory, tokenSource.Token);
+#elif NETSTANDARD2_0
+                    if (!MemoryMarshal.TryGetArray<byte>(_commitHead.AvailableMemory, out var arraySegment))
+                    {
+                        throw new InvalidCastException("Could not get byte[] from Memory.");
```

Comment:
The proper way to do this is to allocate a buffer from the pool and copy it afterwards.

Reply:
Why is that "more proper?"